sql/ttl: Add benchmark to compare timings of TTL expressions

spilchen opened the pull request:
In the process of investigating a TTL-related issue, I developed a microbenchmark to compare the performance of the previous and current TTL expression implementations. The enhancement introduced in PR #102974 allows TTL expressions to utilize the TIMESTAMPTZ + INTERVAL syntax. This microbenchmark aims to simulate and assess the performance impact of that change.

```
dev bench --bench-mem=false ./pkg/sql/ttl/ttljob --filter BenchmarkTTLExpiration --count 10 | tee ~/tmp/new.txt
benchstat -col /query ~/tmp/new.txt

goos: darwin
goarch: arm64
              │    tz_conv    │             interval_math              │
              │    sec/op     │     sec/op      vs base                │
TTLExpiration   0.15455n ± 1%   0.09081n ± 11%  -41.24% (p=0.000 n=10)
```

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141439)
<!-- Reviewable:end -->

fqazi reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141439)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_query_builder_test.go` line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141439#-OJ-evRWDnMM7nOw2-gp:-OJ-evRWDnMM7nOw2-gq:b-plgcys) ([raw file](https://github.com/cockroachdb/cockroach/blob/abe7f0b483b61a90a4f51d646e3a50991a287953/pkg/sql/ttl/ttljob/ttljob_query_builder_test.go#L503)):*
> ```Go
> 	defer srv.Stopper().Stop(ctx)
> 
> 	db := srv.SystemLayer().InternalDB().(descs.DB)
> ```

Think you want ApplicationLayer here instead (that is either the system tenant or default tenant based on the execution mode). 
___
*[`pkg/sql/ttl/ttljob/ttljob_query_builder_test.go` line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141439#-OJ-fXryBZn3KpkEDlYF:-OJ-fXryBZn3KpkEDlYG:b2b7rk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/abe7f0b483b61a90a4f51d646e3a50991a287953/pkg/sql/ttl/ttljob/ttljob_query_builder_test.go#L505)):*
> ```Go
> 	db := srv.SystemLayer().InternalDB().(descs.DB)
> 	err := db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
> 		_, err := txn.Exec(ctx, "create db for benchmark", txn.KV(),
> ```

Nit: Can't we use a normal user connection? 


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/ttl: Add benchmark to compare timings of TTL expressions

In the process of investigating a TTL-related issue, I developed a
microbenchmark to compare the performance of the previous and current
TTL expression implementations. The enhancement introduced in PR #102974
allows TTL expressions to utilize the TIMESTAMPTZ + INTERVAL syntax.
This microbenchmark aims to simulate and assess the performance impact
of that change.

```
dev bench --bench-mem=false ./pkg/sql/ttl/ttljob --filter BenchmarkTTLExpiration --count 10 | tee ~/tmp/new.txt
benchstat -col /query ~/tmp/new.txt

goos: darwin
goarch: arm64
              │    tz_conv    │             interval_math              │
              │    sec/op     │     sec/op      vs base                │
TTLExpiration   0.15455n ± 1%   0.09081n ± 11%  -41.24% (p=0.000 n=10)
```

Epic: none
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141439)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`pkg/sql/ttl/ttljob/ttljob_query_builder_test.go` line 503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141439#-OJ-evRWDnMM7nOw2-gp:-OJ02rif6W5djcmDVQWs:b5qf1ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/abe7f0b483b61a90a4f51d646e3a50991a287953/pkg/sql/ttl/ttljob/ttljob_query_builder_test.go#L503)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Think you want ApplicationLayer here instead (that is either the system tenant or default tenant based on the execution mode).
</blockquote></details>

Yes good point. I opted to just use `serverutils.StartServer` with a sql runner instead.
___
*[`pkg/sql/ttl/ttljob/ttljob_query_builder_test.go` line 505 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141439#-OJ-fXryBZn3KpkEDlYF:-OJ031qT3FmiKZcfxu5v:b543rtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/abe7f0b483b61a90a4f51d646e3a50991a287953/pkg/sql/ttl/ttljob/ttljob_query_builder_test.go#L505)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Can't we use a normal user connection?
</blockquote></details>

I used the sql runner now, which is what I think you meant. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141439#-:-OJ03JlcB5v5GS7aJj5R:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
